### PR TITLE
Con 2259 validation rules batch process

### PR DESCRIPTION
**Pull request by VasanthaKasirajan3008**

### CODE REVIEW

File: src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs [hunk 1]
```diff
@@ -28,14 +28,31 @@ public UpdateCertificatesPrintStatusHandler(ICertificateRepository certificateRe
...
+                if (certificateBatchLog == null)
+                {
+                    //Scenario 3:  Certificate not printed but included in delivery notification file
+                    validationResult.Errors.Add(
+                      new ValidationErrorDetail("CertificatePrintStatuses", $"Certificate {validatedCertificatePrintStatus.CertificateReference} not printed in batch {validatedCertificatePrintStatus.BatchNumber}  .", ValidationStatusCode.NotFound));                    
+                }
+                else
+                {
+                    //Scenario 5:  StatusChangedDateTime earlier than the latest date of the previous status for a certificate number in the given batch for eg delivery date earlier than the printed date
+                    if (validatedCertificatePrintStatus.StatusChangedAt < certificateBatchLog.LatestChange())
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs:46]: This should be certificateBatchLog.StatusAt rather than LatestChange, the difference being when was the status changed rather than when did we know it was changed

File: src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs [hunk 2]
```diff
@@ -28,14 +28,31 @@ public UpdateCertificatesPrintStatusHandler(ICertificateRepository certificateRe
 
         public async Task<ValidationResponse> Handle(UpdateCertificatesPrintStatusRequest request, CancellationToken cancellationToken)
         {
-            var validationResult = new ValidationResponse();
-
+            var validationResult = new ValidationResponse();
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs:31]: You should review your code for 'littering' as well before you commit

File: src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs [hunk 3]
```diff
@@ -70,6 +87,7 @@ private async Task<List<CertificatePrintStatus>> Validate(List<CertificatePrintS
                     .Select(certificatePrintStatus => certificatePrintStatus.Key)
                     .ToList());
 
+            //Scenario 4:  Invalid status in delivery notification file
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs:90]: This looks like you be able to pass 'Printed' in a delivery notification file - what is stopping that - or will you be checking that in the function app as well?

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]
I haven't reviewed the tests yet

File: src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs (hunk 3 above)
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs]: I am checking this in the function app.

File: src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs (hunk 1 above)
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs]: done

File: src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs (hunk 2 above)
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs]: ok , if you chose hide whitespace changes (just for now) you wont see this.

File: src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/BatchLogs/PrintedBatchLogHandlerTests/When_called_and_batch_does_not_exist.cs [hunk 4]
```diff
@@ -14,42 +12,30 @@
...
         public async Task Arrange()
         {
-            MappingBootstrapper.Initialize();
+            //Arrange
+            base.BaseArrange();
 
-            _batchLogQueryRepository = new Mock<IBatchLogQueryRepository>();
-            
+            _batchLogQueryRepository = new Mock<IBatchLogQueryRepository>();            
             _batchLogQueryRepository.Setup(r => r.GetForBatchNumber(It.IsIn(_batchNumber))).Returns(Task.FromResult(_batchLog));
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/BatchLogs/PrintedBatchLogHandlerTests/When_called_and_batch_does_not_exist.cs:27]: This should be able to be moved into the new base in some form

File: src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs [hunk 5]
```diff
@@ -16,26 +16,45 @@ namespace SFA.DAS.AssessorService.Application.Handlers.Certificates
...
+                    {
+                        validationResult.Errors.Add(
+                           new ValidationErrorDetail("StatusChangedDateTime", $"Certificate delivery(StatusChangedAt) datetime {validatedCertificatePrintStatus.StatusChangedAt} earlier than printed(latest date) datetime {certificateBatchLog.LatestChange()}.", ValidationStatusCode.BadRequest));
+                    }
+                }
+
                 var certificate = await _certificateRepository.GetCertificate(validatedCertificatePrintStatus.CertificateReference);
                 if (certificate == null)
-                {
+                {
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs:57]: Unnecessary whitespace change

File: src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs [hunk 6]
```diff
@@ -80,7 +99,7 @@ private async Task<List<CertificatePrintStatus>> Validate(List<CertificatePrintS
             {
                 validationResult.Errors.Add(
                     new ValidationErrorDetail("CertificatePrintStatuses", $"The batch number {invalidBatchNumber} was not found.", ValidationStatusCode.NotFound));
-            });
+            });
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs:102]: And here

File: src/SFA.DAS.AssessorService.Data/CertificateRepository.cs [hunk 7]
```diff
@@ -233,7 +233,7 @@ public async Task<Certificate> GetCertificate(string certificateReference)
                _context.Certificates
                     .FirstOrDefaultAsync(q => q.CertificateReference == certificateReference);
             return certificate;
-        }
+        }
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Data/CertificateRepository.cs:236]: And here

File: src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs [hunk 8]
```diff
@@ -16,26 +16,45 @@ namespace SFA.DAS.AssessorService.Application.Handlers.Certificates
...
+            _certificateBatchLogRepository = certificateBatchLogRepository;
             _mediator = mediator;
             _logger = logger;
         }
 
         public async Task<ValidationResponse> Handle(UpdateCertificatesPrintStatusRequest request, CancellationToken cancellationToken)
         {
             var validationResult = new ValidationResponse();
-
+
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Application/Handlers/Certificates/UpdateCertificatesPrintStatusHandler.cs:36]: And here

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/BatchLogs/PrintedBatchLogHandlerTests/When_called_and_batch_does_not_exist.cs (hunk 4 above)
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/BatchLogs/PrintedBatchLogHandlerTests/When_called_and_batch_does_not_exist.cs]: done

File: src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/BatchLogs/PrintedBatchLogHandlerTests/When_called_and_processed_date_earlier_than_batch_date.cs [hunk 9]
```diff
@@ -15,7 +15,7 @@ namespace SFA.DAS.AssessorService.Application.UnitTests.Handlers.BatchLogs.Print
 {
     public class When_called_and_processed_date_earlier_than_batch_date : PrintedBatchLogHandlerTestBase
     {
-        protected BatchLog _batchLog = new BatchLog { Id = Guid.NewGuid(), BatchNumber = _batchNumber, BatchCreated = DateTime.UtcNow.AddDays(1) };
+        protected new BatchLog _batchLog = new BatchLog { Id = Guid.NewGuid(), BatchNumber = _batchNumber, BatchCreated = DateTime.UtcNow.AddDays(1) };
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/BatchLogs/PrintedBatchLogHandlerTests/When_called_and_processed_date_earlier_than_batch_date.cs:18]: In a test this is not so bad - but that is not a good style, better to have this constructed in a base constructor and then pass the additional value into the base constructor from the derived constructor.

**Review by ChrisJWoodcock — APPROVED** [review]

**Review by shomavg — APPROVED** [review]
